add document about stale read transaction (#6347)
pingcap/docs-cn: stale-read.md

---
title: Stale Read 功能使用场景介绍
aliases: ['/docs-cn/dev/stale-read/']
---

# Stale Read 功能介绍

Stale Read 是 TiDB 提供快速读取数据的一种机制。Stale Read 功能允许用户通过指定时间点或时间范围的方式读取一定程度上的历史数据（使用 [AS OF TIMESTAMP 语法](/as-of-timestamp.md)）。通过 Stale Read，TiDB 可以从任意一个副本上读取到所给定时间点或时间范围内尽可能新的数据，并在这个过程中始终保证数据的一致性约束。

> # 使用场景

Author: updated.

> 在一些跨数据中心部署的场景中，上层应用或业务所请求的大部分事务类型都是只读事务，只存在部分只写事务或读写事务。对于读写事务或者只写事务，业务应用可以容忍一定程度的延迟代价，但是对只读事务来说，其对延迟和 QPS 都有较高的要求。在达到延迟与 QPS 要求的基础上，此类只读事务往往也能够容忍读到一定程度的旧数据，即偏斜的数据（Stale Data）。在 Stale Read 功能中，TiDB 通过牺牲*只读事务*所能读到数据的实效性，并允许从更近距离的副本中读取数据，来在数据跨中心的部署场景中提供更好的读性能。由于 TiDB 使用了 Raft 来作为集群内的副本一致性共识算法，传统的读请求需要由多个副本中的 Leader 角色提供，即便引入了 [Follower Read](/follower-read.md) 功能来减少 Leader 对读请求的处理压力，但为了保证一致性读（Consistent Read）————即在 Follower 上读到的数据在 Leader 上也一定能读到，反之亦然————副本中的 Follower 角色需要在处理读请求之前与 Leader 进行通信，确保自身的数据同步进度至少与 Leader 一样才可以对外提供读请求的处理服务。在跨中心部署的场景下，延迟对上述过程所带来的影响会变得显著，从而降低集群对读请求的处理能力。

Author: updated.

# 使用方式

如前所述，TiDB 提供两种 Stale Read 的方式，分别是指定一个精确的时间点和一个时间范围，两者的区别如下：

- 精确时间点：通过指定一个时间戳，确保读到该指定时间点下保证全局事务记录一致性的数据，不破坏隔离级别，但可能读到旧数据。该功能通过 [AS OF TIMESTAMP 语法](/as-of-timestamp.md#语法方式)提供。
- 时间范围：通过指定一个时间范围，确保读到在该时间范围内尽可能新的数据，不破坏隔离级别，但可能读到旧数据。该功能通过 [AS OF TIMESTAMP 语法](/as-of-timestamp.md#语法方式) 和 [TIDB_BOUNDED_STALENESS 函数](/as-of-timestamp.md#语法方式)实现。

> 除了语法的使用，Stale Read 通常需要配合 TiDB 的 Geo-Partition 部署来进一步在跨中心部署场景中获得更好的性能表现，具体可见 Geo-Partition 使用文档。

Reviewer: 待更新 Geo-Partition 使用文档链接。

Reviewer 2: 先不用提 geo parition 吧，目前没有放出来。而且不限于 geo parition， 多中心部署的都可以用上。

---
title: Stale Read 功能的使用场景
summary: 介绍 Stale Read 功能和使用场景。
---

# Stale Read 功能的使用场景

> 本文档介绍 Stale Read 的使用场景。Stale Read 是 TiDB 用户快速读取数据的一种机制。使用 Stale Read 功能，你能从指定时间点或时间范围内读取一定程度上的历史数据。

Reviewer 3: @nolouch 什么是“一定程度上的历史数据”呢？

Reviewer 2: 不要提 tidb_snapshot 了， 这个用的少，打算废弃掉了。我打算从文档去掉。

> 本文档介绍 Stale Read 的使用场景。由于 TiDB 数据是多版本存储的，Stale Read 能为用户提供了一种读取历史数据的一种机制。使用 Stale Read 功能，你能从指定时间点或时间范围内读取对应的历史数据，从而避免数据同步带来延迟。

Reviewer 3: ```suggestion
本文档介绍 Stale Read 的使用场景。Stale Read 是一种读取历史数据版本的机制，读取 TiDB 中存储的历史数据版本。使用 Stale Read 功能，你能从指定时间点或时间范围内读取对应的历史数据，从而避免数据同步带来延迟。
```

在内部实现上，TiDB 通过 Stale Read 可以从任意一个副本上读取到该指定时间点或时间范围内尽可能新的数据，并在这个过程中始终保证数据的一致性约束。

## 场景描述

+ 场景一：如果一个事务仅涉及只读操作，并且一定程度上可容忍牺牲实时性，你可以使用 Stale Read 功能来读取历史数据。由于牺牲了一定的实时性，使用 Stale Read 后，TiDB 可将请求发送到对应数据的任意一个副本，使得查询的执行获得更大的吞吐量。

> + 场景二：在一些小表的查询场景中，如果使用了强一致性读，数据可能集中在某一个存储节点上，导致查询压力集中在该节点，成为整个查询的瓶颈。使用 Stale Read 功能后，TiDB 可以将请求分发到对应数据的每一个副本上，提升了查询整体的吞吐能力，从而显著提升查询性能。

Reviewer 3: ```suggestion
+ 场景二：在一些小表的查询场景中，如果使用了强一致性读，数据可能集中在某一个存储节点上，导致查询压力集中在该节点，成为整个查询的瓶颈。使用 Stale Read 功能后，TiDB 可以将请求分发到对应数据的每一个副本上，提升查询整体的吞吐能力，从而显著提升查询性能。
```

> 在一些跨数据中心部署的场景中，在使用强一致性读的情况下，可能存在数据请求的源头，与数据实际所在中心不同，从而产生跨数据中心读的情形，导致整体查询的访问延迟上升。通过使用 Stale Read 功能，可以让每个数据请求就近访问对应数据所在当前中心的副本，从而避免跨数据中心读的情形，降低整体查询的访问延迟。

Reviewer 2: 跨数据中心的场景的话需要怎么才能开启就近label 读呢？有相关操作的话也要介绍一下。

> + 场景三：在部分跨数据中心部署的场景中，如果使用了强一致性读，数据请求的源头可能与数据实际所在中心不一致，从而产生跨数据中心读取数据的情况，导致整体查询的访问延迟增加。通过使用 Stale Read 功能，每个数据请求可就近访问对应数据所在当前中心的副本，从而避免跨数据中心读，降低整体查询的访问延迟。

Reviewer 3: ```suggestion
+ 场景三：在部分跨数据中心部署的场景中，如果使用了强一致性读，数据请求的源头可能与实际的数据中心不一致，从而产生跨数据中心读取数据的情况，导致整体查询的访问延迟增加。通过使用 Stale Read 功能，每个数据请求可就近访问对应数据所在当前中心的副本，从而避免跨数据中心读，降低整体查询的访问延迟。
```

> + 场景三：在部分跨数据中心部署的场景中，如果使用了强一致性的 Follower 读，为了读到的数据与 Leader 上的数据一致，会产生跨数据中心获取 Readindex 来校验的请求，导致整体查询的访问延迟增加。通过使用 Stale Read 功能，可以牺牲一定的实时性，可就近访问对应数据所在当前中心的副本，避免跨数据中心的网络延迟，降低整体查询的访问延迟。

Reviewer 2: @Yisaer @JmPotato  这个场景的使用方式如何开启就近读取需要补充一下。

Author: updated.

> + 场景三：在部分跨数据中心部署的场景中，如果使用了强一致性的 Follower 读，为了读到的数据与 Leader 上的数据一致，会产生跨数据中心获取 Readindex 来校验的请求，导致整体查询的访问延迟增加。通过使用 Stale Read 功能，可以牺牲一定的实时性，可就近访问对应数据所在当前中心的副本，避免跨数据中心的网络延迟，降低整体查询的访问延迟。

Reviewer 4: 针对场景三，就近读取到一个尽可能新的数据即可，一定要改 sql，加上「as of timestamp」么，还要明确指定一个时间？

Author: 只要使用 Stale read 功能，就要改 SQL，和哪个场景没关系。

> + 场景三：在部分跨数据中心部署的场景中，如果使用了强一致性的 Follower 读，为了读到的数据与 Leader 上的数据一致，会产生跨数据中心获取 Readindex 来校验的请求，导致整体查询的访问延迟增加。通过使用 Stale Read 功能，可以牺牲一定的实时性，可就近访问对应数据所在当前中心的副本，避免跨数据中心的网络延迟，降低整体查询的访问延迟。

Reviewer 3: ```suggestion
+ 场景三：在部分跨数据中心部署的场景中，如果使用了强一致性的 Follower 读，为了读到的数据与 Leader 上的数据一致，会产生跨数据中心获取 `Readindex` 来校验的请求，导致整体查询的访问延迟增加。通过使用 Stale Read 功能，可以牺牲一定的实时性，可就近访问对应数据所在当前中心的副本，避免跨数据中心的网络延迟，降低整体查询的访问延迟。详情参考[在三数据中心下就近读取数据](/best-practices/three-dc-local-read.md)。
```

## 使用方法

TiDB 提供两种 Stale Read 使用方式，分别是指定一个精确的时间点和一个时间范围：

> - 指定精确时间点：你可以通过指定一个时间戳，确保读到该指定时间点下保证全局事务记录一致性的数据，不破坏隔离级别，但可能读到旧数据。要使用该方法，参阅 [`AS OF TIMESTAMP` 语法](/as-of-timestamp.md#语法方式)文档。

Reviewer 3: ```suggestion
- 指定精确时间点：你可以通过指定一个时间戳，确保读到该指定时间点下保证全局事务记录一致性的数据，不破坏隔离级别，但可能读到旧数据。要使用该方法，参阅 [`AS OF TIMESTAMP` 文档的“语法方式”部分](/as-of-timestamp.md#语法方式)。
```

> - 指定时间范围：你可以通过指定一个时间范围，确保读到在该时间范围内尽可能新的数据，不破坏隔离级别，但可能读到旧数据。要使用该功能，参阅 [`AS OF TIMESTAMP` 语法](/as-of-timestamp.md#语法方式)文档和该文档中 [`TIDB_BOUNDED_STALENESS` 函数](/as-of-timestamp.md#语法方式)部分的介绍。

Reviewer 3: ```suggestion
- 指定时间范围：你可以通过指定一个时间范围，确保读到在该时间范围内尽可能新的数据，不破坏隔离级别，但可能读到旧数据。要使用该功能，参阅 [`AS OF TIMESTAMP` 文档的“语法方式”部分](/as-of-timestamp.md#语法方式)有关 `TIDB_BOUNDED_STALENESS` 函数的介绍。
```

---
title: Stale Read 功能的使用场景
summary: 介绍 Stale Read 功能和使用场景。
---

# Stale Read 功能的使用场景

> 本文档介绍 Stale Read 的使用场景。由于 TiDB 数据是多版本存储的, Stale Read 能为用户提供了一种读取历史数据的一种机制。使用 Stale Read 功能，你能从指定时间点或时间范围内读取对应的历史数据，从而避免数据同步带来延迟。

Reviewer 5: ```suggestion
本文档介绍 Stale Read 的使用场景。由于 TiDB 数据是多版本存储的，Stale Read 能为用户提供了一种读取历史数据的一种机制。使用 Stale Read 功能，你能从指定时间点或时间范围内读取对应的历史数据，从而避免数据同步带来延迟。
```
